File: redundant-connection/redundant-connection.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        int[] par = new int[n+1];
        for(int i=0;i<=n;i++) par[i] = -1;
        int[] ans = new int[2];
        for(int i=0;i<n;i++){
            int x = edges[i][0];
            int y = edges[i][1];
            if(uni(x,y,par)){
                ans[0] = x;
                ans[1] = y;
            }
        }
        return ans;
    }
    
    
    public boolean uni(int x,int y,int[] parent){
        int px = findPar(x,parent);
        int py = findPar(y,parent);
        
        if(px==py){
            return true;
        }
        parent[py] = px;
        return false;
        
    }
    
    public int findPar(int x,int[] parent){
        if(parent[x]==-1) return x;
        return findPar(parent[x],parent);
    }
    
}

File: redundant-connection/redundant-connection.cpp
class Solution {
public:
    int findPar(int node,vector<int>& rank,vector<int>& par){
       if(node==par[node]) return node;
        return par[node] = findPar(par[node],rank,par);
    }
    
    void uni(int a,int b,vector<int>& par,vector<int>& rank){
        a = findPar(a,rank,par);
        b = findPar(b,rank,par);
        if(rank[a]>rank[b]){
            par[b] = a;
        }
        else if(rank[b]>rank[a]){
            par[a] = b;
        }
        else{
            par[b] = a;
            rank[a]++;
        }
    }
    
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size();
        vector<int> par(n,-1);
        vector<int> rank(n,0);
        for(int i=0;i<n;i++) par[i] = i, rank[i] = 0;
        int a = -1,b = -1;
        for(int i=0;i<n;i++){
            int u = edges[i][0];
            int v = edges[i][1];
            u--;
            v--;
            if(findPar(u,rank,par)!=findPar(v,rank,par)){
                uni(u,v,par,rank);
            }
            else{
                a = u;
                b = v;
            }
        }
        vector<int> ans;
        ans.push_back(a+1);
        ans.push_back(b+1);
        return ans;
        
    }
};